### ISSUE TITLE
feat(consensus): propagate foreign proposals within local committees

### ISSUE THREAD

**Issue by sdbondi** [tari-project/tari-dan#595]
## Background

The batching protocol requires that in order to make progress on a decision, a validator must have evidence that any/all involved foreign committees have prepared a transaction.

```text
                       / ---- AllPrepared (foreign LocalPrepared(COMMIT) proposal)
New -> Prepare -> LocalPrepared 
                       \ ---- SomePrepared(at least one foreign LocalPrepared(ABORT) proposal)
```

In the current implementation, the leader broadcasts proposals to all replicas in the local and all foreign committees. 
This has a number of problems:
1. If the leader is Byzantine, they may never broadcast or only broadcast to colluding nodes to gain some advantage
2. The leader may have to connect to an inordinate amount of nodes

The following properties are needed:
1. Each replica local committee should be confident that at least 2f + 1 nodes in each foreign committee have received the proposal
2. A single replica should not be responsible for broadcasting to an inordinate amount of peers
3. Each non-byzantine replica in a foreign committee should be confident that at least 2f + 1 replicas in the same committee have received the proposal

## Implementation

Given a block proposal $B$ containing transactions $tx_1 .. tx_n$ at the LocalPrepared stage involving _distinct_ shards $S = S_1 \bigcup  S_2 .. \bigcup S_n$. 

- A leader proposes block $B$ to all local replicas
- Each non-byzantine replica forwards the proposal to a single node within each foreign shard $S_1, S_2 .. S_n$
- This ensures at least 2f + 1 messages reach each foreign shard and each node only needs to send $|S|$ messages
- Each non-byzantine foreign replica, on receipt of proposal $B$, forwards the proposal to f + 1 local replicas

A component is required that:
- receives foreign proposal messages and determines if it has been seen before (probably need to store the block hash)
- if it has been seen, the message is dropped
- if not, forward to f + 1 local replicas and pass the message onwards to consensus (perhaps this component can mark the transactions as ready without involving the local consensus worker)

**Other considerations**
Block proposals/blocks themselves should have a proposer signature that signs the hash/TransactionId to prevent malleability

**Comment by stringhandler** [comment]
Note that it usually is done by having f+1 local replicas each forward to f+1 foreign replicas, but I believe the 2f+1 strategy also works and results in less messages in most cases

**Comment by sdbondi** [comment]
Yeah, I think if you consider that in practice all replicas will be sending to foreign replicas. Otherwise, how do you know that you're one of the f + 1 (or 2f + 1/non-byzantine) that should send? Having each local send to f + 1 foreign replicas will be more messages than each sending one. We may want to make that a parameter that we can easily change or may have to make it more intelligent based on committee sizes.

**Comment by sdbondi** [comment]
We have an explicit command for foreign proposals that is agreed upon using hotstuff
